pkg/provisioner/ironic/ironic.go

```diff
@@ -1,46 +1,318 @@
+		_, err = nodes.Update(
+			p.client,
+			ironicNode.UUID,
+			nodes.UpdateOpts{
+				nodes.UpdateOperation{
+					Op:    nodes.AddOp,
+					Path:  "/instance_info/image_source",
+					Value: p.instanceImageSource,
+				},
+				nodes.UpdateOperation{
+					Op:    nodes.AddOp,
+					Path:  "/instance_info/image_checksum",
+					Value: p.instanceImageChecksum,
+				},
+			}).Extract()
+		if err != nil {
+			return false, errors.Wrap(err, "failed to update host settings in ironic")
+		}
+
+		// Validate the host settings
+		p.log.Info("validating node settings in ironic")
+		validateResult, err := nodes.Validate(p.client, ironicNode.UUID).Extract()
+		if err != nil {
+			return false, errors.Wrap(err, "failed to validate node settings in ironic")
+		}
+		var validationErrors []string
+		if !validateResult.Boot.Result {
+			validationErrors = append(validationErrors, validateResult.Boot.Reason)
+		}
+		if !validateResult.Deploy.Result {
+			validationErrors = append(validationErrors, validateResult.Deploy.Reason)
+		}
+		if len(validationErrors) > 0 {
+			// We expect to see errors of this nature sometimes, so rather
+			// than reporting it as a reconcile error we record the error
+			// status on the host and return.
+			msg := fmt.Sprintf("host validation error: %s",
+				strings.Join(validationErrors, "; "))
+			dirty = p.host.SetErrorMessage(msg) || dirty
+			return dirty, nil
+		}
+	} else {
+		// FIXME(dhellmann): At this point we have found an existing
+		// node in ironic by looking it up. We need to check its
+		// settings against what we have in the host, and change them
+		// if there are differences.
+	}
+
+	// If we tried to update the node status already and it has an
+	// error, store that value and stop trying to manipulate it.
+	if ironicNode.LastError != "" {
+		// If ironic is reporting an error that probably means it
+		// cannot see the BMC or the credentials are wrong. Set the
+		// error message and return dirty, if we've changed something,
+		// so the status is stored.
+		dirty = p.host.SetErrorMessage(ironicNode.LastError) || dirty
+		return dirty, nil
+	}
+
+	// Ensure the node is marked manageable.
+	//
+	// FIXME(dhellmann): Do we need to check other states here?
+	ironicNode, err = nodes.Get(p.client, p.status.ID).Extract()
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get provisioning state in ironic")
 	}
+	if ironicNode.ProvisionState == string(nodes.Enroll) && ironicNode.TargetProvisionState != nodes.Manageable {
+		p.log.Info("changing provisioning state to manage",
+			"current", ironicNode.ProvisionState,
+			"target", ironicNode.TargetProvisionState,
+		)
+		changeResult := nodes.ChangeProvisionState(
+			p.client,
+			ironicNode.UUID,
+			nodes.ProvisionStateOpts{
+				Target: nodes.TargetManage,
+			})
+		if changeResult.Err != nil {
+			return false, errors.Wrap(changeResult.Err,
+				"failed to change provisioning state to manage")
+		}
+	}
+
```

Comment:
We actually need to hit 'available' state.  Manageable just means there is enough information to manage them.  They then need to be introspected to become 'available' and then we can actually use them.  I see some 'availab'e stuff below but it doesn't look like we are verifying that it's in that state?  Maybe it's just the way the diff is layed out..

Reply:
Good point. For now, I'm just putting the register and delete code up for review. The changes were starting to get big, and I wanted some early validation before going any further.